File: svc-data/src/main/java/com/beancounter/marketdata/providers/alpha/AlphaEnricher.java
package com.beancounter.marketdata.providers.alpha;

import com.beancounter.common.contracts.AssetSearchResponse;
import com.beancounter.common.contracts.AssetSearchResult;
import com.beancounter.common.model.Asset;
import com.beancounter.common.model.Market;
import com.beancounter.marketdata.assets.AssetEnricher;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import lombok.SneakyThrows;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;

@Service
public class AlphaEnricher implements AssetEnricher {
  private final ObjectMapper objectMapper = new AlphaPriceAdapter().getAlphaMapper();
  private final AlphaConfig alphaConfig;
  private AlphaProxyCache alphaProxyCache;
  @Value("${beancounter.market.providers.ALPHA.key:demo}")
  private String apiKey;

  public AlphaEnricher(AlphaConfig alphaConfig) {
    this.alphaConfig = alphaConfig;
  }

  @Autowired(required = false)
  void setAlphaProxyCache(AlphaProxyCache alphaProxyCache) {
    this.alphaProxyCache = alphaProxyCache;
  }

  @SneakyThrows
  @Override
  public Asset enrich(Market market, String code, String name) {
    String marketCode = alphaConfig.translateMarketCode(market);
    String symbol = alphaConfig.translateSymbol(code);
    if (marketCode != null) {
      symbol = symbol + "." + marketCode;
    }
    String result = alphaProxyCache.search(symbol, apiKey).get();
    AssetSearchResult assetResult = getAssetSearchResult(market, result);
    if (assetResult == null) {
      return null;
    }
    return Asset.builder()
        .code(code.toUpperCase())
        .name(assetResult.getName())
        .priceSymbol(assetResult.getSymbol())
        .market(market)
        .marketCode(market.getCode().toUpperCase())
        .category(assetResult.getType())
        .build();
  }

  private AssetSearchResult getAssetSearchResult(Market market, String result)
      throws JsonProcessingException {

    AssetSearchResponse assetSearchResponse =
        objectMapper.readValue(result, AssetSearchResponse.class);
    if (assetSearchResponse.getData() == null || assetSearchResponse.getData().isEmpty()) {
      return null;
    }
    AssetSearchResult assetResult = assetSearchResponse.getData().iterator().next();
    if (!assetResult.getCurrency().equals(market.getCurrencyId())) {
      // Fuzzy search result returned and asset from a different exchange
      return null;
    }
    return assetResult;
  }

  @Override
  public boolean canEnrich(Asset asset) {
    return asset.getName() == null;
  }
}

File: svc-data/src/main/kotlin/com/beancounter/marketdata/providers/alpha/AlphaEnricher.kt
package com.beancounter.marketdata.providers.alpha

import com.beancounter.common.contracts.AssetSearchResponse
import com.beancounter.common.contracts.AssetSearchResult
import com.beancounter.common.exception.SystemException
import com.beancounter.common.model.Asset
import com.beancounter.common.model.Market
import com.beancounter.marketdata.assets.AssetEnricher
import com.fasterxml.jackson.core.JsonProcessingException
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Service
import java.util.Locale

/**
 * Backfills missing asset data from a 3rd party. Basically adds asset.name for a supplied asset.code.
 */
@Service
class AlphaEnricher(private val alphaConfig: AlphaConfig) : AssetEnricher {
    private val objectMapper = AlphaPriceAdapter().alphaMapper
    private var alphaProxyCache: AlphaProxyCache? = null

    @Value("\${beancounter.market.providers.ALPHA.key:demo}")
    private val apiKey: String? = null

    @Autowired(required = false)
    fun setAlphaProxyCache(alphaProxyCache: AlphaProxyCache?) {
        this.alphaProxyCache = alphaProxyCache
    }

    override fun enrich(id: String, market: Market, code: String, defaultName: String?): Asset? {
        val marketCode = alphaConfig.translateMarketCode(market)
        var symbol = alphaConfig.translateSymbol(code)
        if (marketCode != null) {
            symbol = "$symbol.$marketCode"
        }
        val result = alphaProxyCache!!.search(symbol, apiKey)
        // var assetResult: AssetSearchResult? = null
        val assetResult = try {
            getAssetSearchResult(market, result)
        } catch (e: JsonProcessingException) {
            throw SystemException("This shouldn't have happened")
        }
        return if (assetResult == null) {
            null
        } else Asset(
            id,
            code.uppercase(Locale.getDefault()),
            name = assetResult.name,
            category = assetResult.type,
            market = market,
            marketCode = market.code,
            priceSymbol = assetResult.symbol!!,
        )
    }

    @Throws(JsonProcessingException::class)
    private fun getAssetSearchResult(market: Market, result: String?): AssetSearchResult? {
        val (data) = objectMapper.readValue(result, AssetSearchResponse::class.java)
        if (data.isEmpty()) {
            return null
        }
        val assetResult = data.iterator().next()
        return if (currencyMatch(assetResult.currency, market.currencyId)) {
            assetResult
        } else
        // Fuzzy search result returned and asset from a different exchange
            null
    }

    fun currencyMatch(currency: String?, currencyId: String): Boolean {
        var match = currency
        if (currency == "GBX") {
            match = "GBP"
        }
        return match == currencyId
    }

    override fun canEnrich(asset: Asset): Boolean {
        return asset.name == null
    }

    override fun id(): String {
        return "ALPHA"
    }
}
